### PR TITLE
Pass enabled and disabled extensions to apply-refact

### PR DESCRIPTION
**Pull request by zliu41**
Related: #504

For testing, enabled = defaultExtensions, disabled = [].

This uncovered a bug in apply-refact, see https://github.com/mpickering/apply-refact/issues/58

### CODE REVIEW

File: src/HLint.hs [hunk 1]
```diff
@@ -184,8 +186,8 @@ getIdeas cmd@CmdMain{..} settings = do
 
 -- #746: run refactor even if no hint, which ensures consistent output
 -- whether there are hints or not.
-handleRefactoring :: [Idea] -> [String] -> Cmd -> IO ()
-handleRefactoring ideas files cmd@CmdMain{..} =
+handleRefactoring :: [Idea] -> [String] -> Cmd -> [Extension] -> [Extension] -> IO ()
+handleRefactoring ideas files cmd@CmdMain{..} ys ns =
```
Comment (ndmitchell) [inline src/HLint.hs:190]: Why not have handleRefactoring just do:

```
let ParseFlags{enabledExtensions, disabledExtensions} = cmdParseFlags cmd
```

Everything else just passes around the `CmdMain` flags and generates the parse flags from it when they need it. I appreciate that results in calling `cmdParseFlags` twice (more expensive), but that seems better than relying on the invariant that the flags match the command.

File: src/HLint.hs [hunk 2]
```diff
@@ -163,18 +166,17 @@ runHints args settings cmd@CmdMain{..} = do
...
             usecolour <- cmdUseColour cmd
             showItem <- if usecolour then showANSI else pure show
             mapM_ (outStrLn . showItem) ideas
             handleReporting ideas cmd
         pure ideas
 
-getIdeas :: Cmd -> [Setting] -> IO [Idea]
-getIdeas cmd@CmdMain{..} settings = do
+getIdeas :: Cmd -> ParseFlags -> [Setting] -> IO [Idea]
+getIdeas CmdMain{..} flags settings = do
```
Comment (ndmitchell) [inline src/HLint.hs:178]: I wouldn't bother passing flags here. Recomputing them is nice and easy.

File: src/Refact.hs [hunk 3]
```diff
@@ -53,9 +54,10 @@ refactorPath rpath = do
                        , "<https://github.com/mpickering/apply-refact>"
                        ]
 
-runRefactoring :: FilePath -> FilePath -> FilePath -> String -> IO ExitCode
-runRefactoring rpath fin hints opts =  do
+runRefactoring :: FilePath -> FilePath -> FilePath -> [Extension] -> [Extension] -> String -> IO ExitCode
+runRefactoring rpath fin hints ys ns opts =  do
```
Comment (ndmitchell) [inline src/Refact.hs:58]: Perhaps `enabled disabled` as the names? I appreciate that `ys`/`ns` was meant to signify that, but it wasn't immediately obvious to me.

File: src/Extension.hs [hunk 4]
```diff
@@ -24,6 +24,9 @@ reallyBadExtensions =
   {- , XmlSyntax , RegularPatterns -} -- steals a-b and < operators
   , AlternativeLayoutRule -- Does not play well with 'MultiWayIf'
   , NegativeLiterals -- Was not enabled by HSE and enabling breaks tests.
+  , StarIsType -- conflicts with TypeOperators. StarIsType is currently enabled by default,
+               -- so adding it here has no effect except avoiding passing it to apply-refact.
+               -- See https://github.com/mpickering/apply-refact/issues/58
```
Comment (ndmitchell) [inline src/Extension.hs:29]: Great detective work!

**Comment by ndmitchell** [comment]
Amazing work!

**Comment by zliu41** [comment]
Thanks!

**Pull request merged**